### PR TITLE
Use dependency injection for proc memory mocks

### PR DESCRIPTION
**Pull request by fjetter**
The mocks introduced in https://github.com/dask/distributed/pull/5878 are unstable in tests like [`test_fail_to_pickle_spill`](https://github.com/dask/distributed/blob/d667a9055169253b55fc079fe72c14aabf7ef2fc/distributed/tests/test_worker_memory.py#L141-L152)
because the mocks only register once the test is entered. With the extremely small memory limit, and the extreme small monitor-interval, the event loop appears to be stressed out sufficiently that the workers actually never come up in time and the test times out. Increasing the interval is one option. I encountered this during https://github.com/dask/distributed/pull/5910

The mocks are not to blame here but rather that we're patching an initialized instance after it has been started. IMO, the proper way would be to use dependency injection using a fake object instead of a monkeypatch.

This introduces a dependency injection pattern for the WorkerMemoryManager and defines a specific mock target as motivated already in https://github.com/dask/distributed/pull/5870#issuecomment-1054165512

### CODE REVIEW

File: distributed/worker_memory.py [hunk 1]
```diff
@@ -302,32 +313,46 @@ def __init__(
...
-            memory = proc.memory_info().rss
-        except (ProcessLookupError, psutil.NoSuchProcess, psutil.AccessDenied):
+        nanny = self.nanny()
+        if (
+            not nanny
+            or nanny.status != Status.running
+            or nanny.process is None
+            or nanny.process.process is None
+            or self.memory_limit is None
+        ):
```
Comment (crusaderky) [inline distributed/worker_memory.py:347]:
```suggestion
        ):        
            assert self.memory_limit is not None
            assert self.memory_terminate_fraction is not False
```

File: distributed/worker_memory.py [hunk 2]
```diff
@@ -302,32 +313,46 @@ def __init__(
             dask.config.get("distributed.worker.memory.monitor-interval"),
             default=None,
         )
+        self.nanny = weakref.ref(nanny)
```
Comment (crusaderky) [inline distributed/worker_memory.py:316]: This is self._worker above - could you make it consistent?
Also, please add to the attributes declarations above in both classes.

**Review by crusaderky — COMMENTED** [review]
IMHO, the whole thing feels over-designed here.
The failures (for reference: https://github.com/dask/distributed/runs/5693287345?check_suite_focus=true) are both caused by 
1. having ``"memory_limit": "1 GB"``
2. the worker is already beyond 1GB RAM when the worker starts, as it runs in the main process
3. the test is about pause, so the pause kicks in before the monkey patch of the instance can happen.

The solution is very simple - just make memory_limit and all mocked measures safely larger than what the whole test suite will ever consume - e.g. 10 GB.

File: distributed/worker_memory.py (hunk 1 above)
Comment (fjetter) [reply distributed/worker_memory.py:347]: A few lines below we're calculating `memory / self.memory_limit`, i.e. if memory_limit is None we can skip `memory_monitor` / return early.


So, why remove the `self.memory_limit is None` from the if clause and why add these asserts? If these asserts are a concerning edge case, we should probably rather add a unit test

**Comment by fjetter** [comment]
> IMHO, the whole thing feels over-designed here.

I understand this concern. I still thing we should go through with this. I don't think mocking something mid flight is a stable way of testing. The fact that the pause may already kick in before the test even starts, regardless of how the settings are configured is not stable and harder to maintain. Even if setting the parameters properly avoids this issue, there is a lot of knowledge required to A) identify the situation and B) to set the proper values. Instead of a monkeypatch, a fake object, i.e. a fully functional object with a few simplifications (e.g. the process memory) is typically a more robust design and this is what I am proposing here.

Providing the type of `WorkerMemoryManager` to the worker appears to be the simplest way to achieve this since the factoring out of the memory related code. If there is another method to achieve this, I'm open to exploring but I really would like to have our patch/fake in place before the server starts.

File: distributed/worker_memory.py (hunk 1 above)
Comment (crusaderky) [reply distributed/worker_memory.py:349]: it's not an edge case; the whole method will never be scheduled if memory_limit is disabled. See lines ~140 in ``__init__``. There are already several tests for it.

**Comment by crusaderky** [comment]
> Even if setting the parameters properly avoids this issue, there is a lot of knowledge required to A) identify the situation and B) to set the proper values.

The log of the failed tests (https://github.com/dask/distributed/runs/5693287345?check_suite_focus=true) states:
```
2022-03-25 14:54:22,158 - distributed.worker_memory - WARNING - Worker is at 102% memory usage. Pausing worker.  Process memory: 0.95 GiB -- Worker memory limit: 0.93 GiB
2022-03-25 14:54:51,752 - distributed.utils_test - ERROR - Failed to start gen_cluster: TimeoutError: Cluster creation timeout; retrying
```
it looks very explicit to me.

> regardless of how the settings are configured is not stable and harder to maintain. 

I think it's a safe statement that the main process of our test suite will never reach 10GB permanent RAM. You can use 100GB or 1TB if you want to remove all ambiguity.

> The fact that the pause may already kick in before the test even starts, regardless of how the settings are configured 

The pause may kick in on *any* test decorated by ``@gen_cluster`` if enough memory is leaked by previous tests.

> a fully functional object with a few simplifications (e.g. the process memory) is typically a more robust design and this is what I am proposing here.

I would agree for the actual production code. I disagree for unit tests - you're just making them harder to maintain.

> Providing the type of WorkerMemoryManager to the worker appears to be the simplest way to achieve this since the factoring out of the memory related code. If there is another method to achieve this, I'm open to exploring but I really would like to have our patch/fake in place before the server starts.

You're adding complexity to Worker for the only purpose of unit testing. This to me is an antipattern.

**Comment by fjetter** [comment]
Closing in favour of https://github.com/dask/distributed/pull/6055
